NatNumberCategories

```diff
@@ -0,0 +1,416 @@
+  dsimp [Nat.Perfect]
+  rw [show properDivisors 6 = {1, 2, 3} by rfl]
+  norm_num
+
+theorem twelve_abundant : Nat.Abundant 12 := by
+  dsimp [Nat.Abundant]
+  rw [show properDivisors 12 = {1,2,3,4,6} by rfl]
+  norm_num
+
+theorem seventy_weird : Nat.WeirdNumber 70 := by
+  dsimp [Nat.WeirdNumber, Nat.Abundant]
+  rw [not_semiperfect_iff_forall]
+  have div70 : properDivisors 70 = {1, 2, 5, 7, 10, 14, 35} := by rfl
+  constructor
+  · constructor
+    rw [div70]
+    repeat norm_num
+  · rw [div70]
+    intro s hs
+    left
+    fin_cases hs <;> decide
+
+/- ### END OF SPECIFIC EXAMPLES ### -/
+
+theorem divisor_subseteq_mul (hn : 0 < n) (hm : 0 < m) :
+  Nat.divisors n ∪ Nat.divisors m ⊆ Nat.divisors (n*m) := by
+  have nm_ne_zero : n*m ≠ 0 := by
+    simp_all only [ne_eq, _root_.mul_eq_zero]
+    rw [not_or]
+    constructor
+    exact Nat.pos_iff_ne_zero.mp hn
+    exact Nat.pos_iff_ne_zero.mp hm
+  refine subset_iff.mpr ?_
+  intro a ha
+  have ha' : a ∈ divisors n ∨ a ∈ divisors m := by
+    simp_all only [mem_union, mem_divisors, ne_eq]
+  rcases ha' with hn1 | hm1
+  · have han : a ∣ n := by
+      simp_all only [mem_union, mem_divisors, ne_eq]
+    refine mem_divisors.mpr ?inl.a
+    constructor
+    exact Nat.dvd_trans han (Exists.intro m rfl)
+    assumption
+  · have ham : a ∣ m := by
+      simp_all only [mem_union, mem_divisors, ne_eq]
+    refine mem_divisors.mpr ?inr.a
+    constructor
+    exact Dvd.dvd.mul_left ham n
+    assumption
+
+theorem prop_divisors_subseteq_mul (hn : 0 < n) (hm : 0 < m) :
+  Nat.properDivisors n ∪ Nat.properDivisors m ⊆ Nat.properDivisors (n*m) := by
+  refine subset_iff.mpr ?_
+  rintro a ha
+  have ha' : a ∈ properDivisors n ∨ a ∈ properDivisors m := by
+    simp_all only [gt_iff_lt, le_mul_iff_one_le_right,
+      le_mul_iff_one_le_left, mem_union, mem_properDivisors]
+  rcases ha' with hn1 | hm1
+  · have han : a ∣ n := by
+      simp_all only [mem_properDivisors, and_self]
+    refine mem_properDivisors.mpr ?_
+    constructor
+    exact Dvd.dvd.mul_right han m
+    refine lt_mul_of_lt_of_one_le ?_ hm
+    simp_all only [gt_iff_lt, le_mul_iff_one_le_right, mem_properDivisors, true_and, ne_eq]
+  · have ham : a ∣ m := by
+      simp_all only [gt_iff_lt, le_mul_iff_one_le_right, mem_properDivisors]
+    refine mem_properDivisors.mpr ?_
+    constructor
+    exact Dvd.dvd.mul_left ham n
+    refine lt_mul_of_one_le_of_lt hn ?_
+    simp_all only [gt_iff_lt, le_mul_iff_one_le_right, le_mul_iff_one_le_left, mem_properDivisors,
+       true_and, ne_eq]
+
+lemma divisors_eq_proper_union_self (hn : 0 < n) :
+  Nat.divisors n = Nat.properDivisors n ∪ {n} := by
+    dsimp [divisors, properDivisors]
+    ext a; simp [*]
+    constructor
+    rintro ⟨⟨h1, h2⟩, h3⟩
+    by_cases hn : a ≠ n
+    left
+    repeat constructor
+    · assumption
+    · refine (LE.le.gt_iff_ne ?_).mpr (id (Ne.symm hn))
+      exact lt_succ.mp h2
+    · exact h3
+    repeat tauto
+    aesop
+    linarith
```

Comment:
This proof is a mess. You have a non-terminal `simp`, you have an `aesop` call which fails (and it prints a message to stdout to tell you so), a whole load of `repeats` etc – this is the sort of thing which first-time Lean coders produce, throwing every conceivable tactic at the goal until something somehow sticks. 

I'm afraid I'm going to have to be brutal here: this proof shows that you do not yet have the skills needed to contribute effectively to mathlib, and that continuing with reviewing this PR is not a good use of my time. I recommend you work on your Lean coding skills, study code written by other more experienced coders, and hold off from attempting to contribute to mathlib until you are in a position to do this effectively.

FWIW, if I were trying to prove this statement, I'd probably produce something like
````
lemma divisors_eq_proper_union_self (hn : 0 < n) :
    divisors n = properDivisors n ∪ {n} := by
  ext1 a
  rw [mem_divisors, mem_union, mem_properDivisors, mem_singleton]
  constructor
  · intro ⟨hdiv, hlt⟩
    rcases eq_or_lt_of_le (le_of_dvd hn hdiv) with rfl | hlt' <;> tauto
  · rintro (⟨hdiv, _⟩ | rfl)
    · exact ⟨hdiv, hn.ne'⟩
    · exact ⟨dvd_rfl, hn.ne'⟩
````

Reply:
Let's judge the proof, not the contributor. Many contributors to mathlib start without knowing what makes a good contribution to mathlib, and the tough job of a reviewer is help contributors rewrite their PRs to become acceptable. Pointing out non-terminal `simp`s and unnecessary tactics is good. Providing a rewritten proof is good. But I think the other elements are nonconstructive and not something that is necessary to say.

Reply:
No worries. I am pretty new to this Mathlib stuff so I appreciate the criticism as it helps me learn. I am currently working on the file on my computer and fixing it up. When it's done I'll reopen the pull request. Thank you to @loefflerd for the suggestions and @kmill for the encouragement.

Reply:
Also I've been busy with university and research so progress has been a bit slow. But, I plan to get back into it soon.